password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,164 @@
+Stability   : experimental
+Portability : POSIX
+
+= validate
+
+It is common for passwords to have a set of requirements.
+(e.g. Password must contain at least eight characters that consist of alphabetic
+characters combined with numbers or special characters.)
+
+Validate module provides set of functions which enables you to validate them.
+
+-}
+
+module Data.Password.Validate
+  ( PasswordPolicy (..),
+    PolicyCharSet(..),
+    FailedPolicies (..),
+    CharacterCategory(..),
+    defaultPasswordPolicy,
+    valid,
+    validatePassword,
+    specialChars,
+  ) where
+
+import Data.Char (isAscii, isDigit, isLower, isUpper)
+import Data.Maybe (catMaybes)
+import Data.Password.Internal (Password (..))
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Charset
+data PolicyCharSet = PolicyCharSetAscii
+  deriving (Eq, Ord, Show)
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy
+  = PasswordPolicy
+      { passPolicyMinimumLength :: !Int,
+        -- ^ Required password minimum length
+        passPolicyMaximumLength :: !Int,
+        -- ^ Required password maximum length
+        passPolicyCharUppercase :: !(Maybe Int),
+        -- ^ Required number of upper-case characters
+        passPolicyCharLowercase :: !(Maybe Int),
+        -- ^ Required number of lower-case characters
+        passPolicyCharSpecial :: !(Maybe Int),
+        -- ^ Required number of special characters
+        passPolicyCharDigit :: !(Maybe Int),
+        -- ^ Required number of ASCII-digit characters
+        passPolicyCharSet :: PolicyCharSet
+        -- ^ Set of characters that can be used for password
+      } deriving (Eq, Ord, Show)
+
+-- | Default value for the 'PasswordPolicy'
+--
+-- >>> defaultPasswordPolicy
+-- PasswordPolicy {passPolicyMinimumLength = 8, passPolicyMaximumLength = 30, passPolicyCharUppercase = Just 1, passPolicyCharLowercase = Just 1, passPolicyCharSpecial = Nothing, passPolicyCharDigit = Just 1, passPolicyCharSet = PolicyCharSetAscii}
+defaultPasswordPolicy :: PasswordPolicy
+defaultPasswordPolicy = PasswordPolicy
+  { passPolicyMinimumLength = 8,
+    passPolicyMaximumLength = 30,
+    passPolicyCharUppercase = Just 1,
+    passPolicyCharLowercase = Just 1,
+    passPolicyCharSpecial = Nothing,
+    passPolicyCharDigit = Just 1,
+    passPolicyCharSet = PolicyCharSetAscii
+  }
+
+-- | Character Category
+data CharacterCategory
+  = Uppercase
+  | Lowercase
+  | Special
+  | Digit
+  deriving (Eq, Ord, Show)
+
+-- | Data type representing how password are invalid
+data FailedPolicies
+  = PasswordTooShort Int Int
+  -- ^ Password is too short
+  | PasswordTooLong Int Int
+  -- ^ Password is too long
+  | NotEnoughReqChars CharacterCategory Int Int
+  -- ^ Password does not contain characters that are required
+  | InvalidChar PolicyCharSet
+  -- ^ Password contains invalid character sets
+  deriving (Eq, Ord, Show)
```

Comment:
Validation could return false due to various reasons. In order for us to properly test it, I've introduces
`FailedPolicies` type. This allows us to properly test the validation using property-based testing such as `QuickCheck`.

Reply:
Having a failure sum type will also be nice to return if the validation fails,
(e.g. `[FailedPolicies]`) so that all the reasons the validation fails can be communicated back to the user in some way.

Reply:
It might also be nice to have a custom `Show` instance, so you automatically get a nice message when `show`-ing a `PolicyFailure`.

Reply:
@Vlix 
I decided to name it `InvalidReason`, referring to the [X509](https://hackage.haskell.org/package/x509-validation-1.6.11/docs/Data-X509-Validation.html#v:validate) package.

Regarding having a custom `Show` instance, I personally thing it's unnecessary.
If `password` is an application, it should implement a polite error message, but since this is a package, I think it's up to the developers to implement their own rendering function (`PolicyFailure -> Text`).